xen/arch/arm/coproc/plat/coproc_xxx.c

```diff
@@ -0,0 +1,275 @@
+/*
+ * xen/arch/arm/coproc/plat/coproc_xxx.c
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+
+#include <xen/device_tree.h>
+#include <xen/errno.h>
+#include <xen/err.h>
+#include <xen/irq.h>
+#include <xen/list.h>
+#include <xen/vmap.h>
+#include <asm/device.h>
+#include <asm/types.h>
+#include <asm/platform.h>
+
+#include "../coproc.h"
+#include "coproc_xxx.h"
+
+#define DT_MATCH_COPROC_XXX DT_MATCH_COMPATIBLE("vendor_xxx,coproc_xxx")
+
+static int vcoproc_xxx_read(struct vcpu *v, mmio_info_t *info, register_t *r, void *priv)
+{
+	struct vcoproc_info *vcoproc_xxx = priv;
+
+	(void)vcoproc_xxx;
+
+	return 0;
+}
+
+static int vcoproc_xxx_write(struct vcpu *v, mmio_info_t *info, register_t r, void *priv)
+{
+	struct vcoproc_info *vcoproc_xxx = priv;
+
+	(void)vcoproc_xxx;
+
+	return 0;
+}
+
+static const struct mmio_handler_ops vcoproc_xxx_mmio_handler = {
+	.read = vcoproc_xxx_read,
+	.write = vcoproc_xxx_write,
+};
+
+
+static int vcoproc_xxx_domain_init(struct domain *d, struct vcoproc_info *vcoproc_xxx)
+{
+	struct coproc_device *coproc_xxx = vcoproc_xxx->coproc;
+	int i;
+
+	for (i = 0; i < coproc_xxx->num_mmios; i++) {
+		struct mmio *mmio = &coproc_xxx->mmios[i];
+
+		register_mmio_handler(d, &vcoproc_xxx_mmio_handler,
+				mmio->addr, mmio->size, vcoproc_xxx);
+	}
+
+	return 0;
+}
+
+static void vcoproc_xxx_domain_free(struct domain *d, struct vcoproc_info *vcoproc_xxx)
+{
+
+}
+
+static const struct vcoproc_domain_ops vcoproc_xxx_domain_ops = {
+	.domain_init = vcoproc_xxx_domain_init,
+	.domain_free = vcoproc_xxx_domain_free,
+};
+
+static int vcoproc_xxx_vcoproc_init(struct domain *d, struct coproc_device *coproc_xxx)
+{
+	struct vcoproc_info *vcoproc_xxx;
+	int ret;
+
+	vcoproc_xxx = xzalloc(struct vcoproc_info);
+	if (!vcoproc_xxx) {
+		dev_err(coproc_xxx->dev, "failed to allocate vcoproc_info\n");
+		return -ENOMEM;
+	}
+
+	vcoproc_xxx->coproc = coproc_xxx;
+	vcoproc_xxx->domain = d;
+	vcoproc_xxx->ops = &vcoproc_xxx_domain_ops;
+	spin_lock_init(&vcoproc_xxx->lock);
+
+	spin_lock(&coproc_xxx->vcoprocs_lock);
+	list_add(&vcoproc_xxx->list, &coproc_xxx->vcoprocs_list);
+	spin_unlock(&coproc_xxx->vcoprocs_lock);
+
+	ret = vcoproc_attach(d, vcoproc_xxx);
+	if (ret) {
+		dev_err(coproc_xxx->dev, "failed to attach vcoproc to domain %d\n", d->domain_id);
+		return ret;
+	}
```

Comment:
first attach the vcoproc to the domain then register vcoproc in coproc framework
or should take care to cleanup registration on failed attachement